src/lib/crypto/kimchi_backend/gadgets/keccak.ml

```diff
@@ -0,0 +1,720 @@
+          (* Field element containing value 2^(8*z) *)
+          let shift_field =
+            Common.bignum_bigint_to_field
+              (module Circuit)
+              Bignum_bigint.(pow (of_int 2) (of_int (Int.( * ) 8 z)))
+          in
+          let shift = Field.constant shift_field in
+          (* TODO: Does this generate automatic generic gates to check composition? *)
+          Field.Assert.equal (Field.constant shift_field) shift ;
+          state.(x).(y) <-
+            Field.(state.(x).(y) + (shift * List.nth_exn bytestring index))
+        done
+      done
+    done ;
+    state
+
+  (* Converts a state of cvars to a list of bytes as cvars *)
+  let as_prover_to_bytes (type f)
+      (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+      (state : Circuit.Field.t matrix) : Circuit.Field.t list =
+    let open Circuit in
+    assert (
+      Array.length state = keccak_dim && Array.length state.(0) = keccak_dim ) ;
+    let bytestring = Array.create ~len:200 Field.zero in
+    let bytes_per_word = keccak_word / 8 in
+    for y = 0 to keccak_dim - 1 do
+      for x = 0 to keccak_dim - 1 do
+        let base_index = bytes_per_word * ((keccak_dim * y) + x) in
+        for z = 0 to bytes_per_word - 1 do
+          let index = base_index + z in
+          let byte =
+            exists Field.typ ~compute:(fun () ->
+                let word =
+                  Common.cvar_field_to_bignum_bigint_as_prover
+                    (module Circuit)
+                    state.(x).(y)
+                in
+                let power_lo = 8 * z in
+                let offset_lo =
+                  Bignum_bigint.(pow (of_int 2) (of_int power_lo))
+                in
+                let two_pow_8 = Bignum_bigint.(pow (of_int 2) (of_int 8)) in
+                let byte =
+                  Bignum_bigint.(
+                    (word - (word % offset_lo)) / offset_lo % two_pow_8)
+                in
+                Common.bignum_bigint_to_field (module Circuit) byte )
+          in
+          bytestring.(index) <- byte
+        done ;
+
+        (* TODO: Does this generate automatic generic gates to check decomposition? *)
+
+        (* Create a list containing the elements of bytestring from base_index to base_index + 7 *)
+        let word_bytes =
+          Array.to_list
+          @@ Array.sub bytestring ~pos:base_index ~len:bytes_per_word
+        in
+
+        (* Assert correct decomposition of bytes from state *)
+        check_bytes_to_word (module Circuit) state.(x).(y) word_bytes
+      done
+    done ;
+    Array.to_list bytestring
+
+  let xor (type f)
+      (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+      (input1 : Circuit.Field.t matrix) (input2 : Circuit.Field.t matrix) :
+      Circuit.Field.t matrix =
+    let open Circuit in
+    assert (
+      Array.length input1 = keccak_dim && Array.length input1.(0) = keccak_dim ) ;
```

Comment:
Same as above, I just had them so that a future circuit writer does not misuse it. Those are just normal asserts, they don't add rows to the circuit. I believed it was just a cheap sanity check. But I can remove it and instead add a disclaimer in front of the function to explain how it should be called.